### PR TITLE
[ROCm] Retry loop implemented to avoid transient memory leak errors

### PR DESCRIPTION
**Pull request by alugorey**
### Description
Added a retry loop to memory leak checker to avoid rare case in which ROCM reports a false positive memory leak.

### Issue
Original issue observed as part of this ticket: https://github.com/pytorch/pytorch/issues/62533

### Testing
- Applied changes and built
- python test/test_cuda.py
- Ensure all tests pass

### CODE REVIEW

**Review by malfet — APPROVED** [review]

**Review by malfet — COMMENTED** [review]
Looks fine to me, but it would be nice to abort the loop on the first run if discrepancy is not detected

**Comment by malfet** [comment]
Also, @ZainRizvi once this is landed let's monitor if some of the test becomes more flaky

File: torch/testing/_internal/common_utils.py [hunk 1]
```diff
@@ -1487,18 +1487,31 @@ def __exit__(self, exec_type, exec_value, traceback):
...
+                if caching_allocator_mem_allocated > self.caching_allocator_befores[i]:
+                    caching_allocator_discrepancy = True
+
+                if driver_mem_allocated > self.driver_befores[i]:
+                    driver_discrepancy = True
+
+                if not(caching_allocator_discrepancy or driver_discrepancy):
+                    # Leak was false positive, exit loop
+                    discrepancy_detected = False
+                    break
```
Comment (jithunnair-amd) [inline torch/testing/_internal/common_utils.py:1511]: Should it not be the following?
```suggestion
                    discrepancy_detected = False
                
                if discrepancy_detected = False:
                    break
```

Comment (jithunnair-amd) [reply torch/testing/_internal/common_utils.py:1511]: Sorry, looks like these should be functionally same.

**Comment by jithunnair-amd** [comment]
ROCm and CUDA CI are clean. 4 CI failures are unrelated.

@pytorchbot merge

**Pull request closed without merging**